server/sonar-server/src/main/java/org/sonar/server/issue/ws/BulkChangeAction.java

```diff
@@ -0,0 +1,386 @@
+      .setDeprecatedSince("5.5");
+    action.createParam(PARAM_DO_TRANSITION)
+      .setDescription("Transition")
+      .setExampleValue(REOPEN)
+      .setPossibleValues(DefaultTransitions.ALL)
+      .setDeprecatedKey("do_transition.transition");
+    action.createParam(PARAM_ADD_TAGS)
+      .setDescription("Add tags")
+      .setExampleValue("security,java8");
+    action.createParam(PARAM_REMOVE_TAGS)
+      .setDescription("Remove tags")
+      .setExampleValue("security,java8");
+    action.createParam(PARAM_COMMENT)
+      .setDescription("To add a comment to a list of issues")
+      .setExampleValue("Here is my comment");
+    action.createParam(PARAM_SEND_NOTIFICATIONS)
+      .setDescription("Available since version 4.0")
+      .setBooleanPossibleValues()
+      .setDefaultValue("false");
+  }
+
+  @Override
+  public void handle(Request request, Response response) throws Exception {
+    userSession.checkLoggedIn();
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      Issues.BulkChangeWsResponse wsResponse = Stream.of(request)
+        .map(loadData(dbSession))
+        .map(executeBulkChange())
+        .map(toWsResponse())
+        .collect(Collectors.toOneElement());
+      writeProtobuf(wsResponse, request, response);
+    }
+  }
+
+  private Function<Request, BulkChangeData> loadData(DbSession dbSession) {
+    return request -> new BulkChangeData(dbSession, request);
+  }
+
+  private Function<BulkChangeData, BulkChangeResult> executeBulkChange() {
+    return bulkChangeData -> {
+      BulkChangeResult result = new BulkChangeResult(bulkChangeData.issues.size());
+      IssueChangeContext issueChangeContext = IssueChangeContext.createUser(new Date(system2.now()), userSession.getLogin());
+      bulkChangeData.issues.stream()
+        .filter(bulkChange(issueChangeContext, bulkChangeData, result))
+        .peek(issueStorage::save)
+        .forEach(sendNotification(issueChangeContext, bulkChangeData));
+      return result;
+    };
+  }
+
+  private Predicate<DefaultIssue> bulkChange(IssueChangeContext issueChangeContext, BulkChangeData bulkChangeData, BulkChangeResult result) {
+    return issue -> {
+      ActionContext actionContext = new ActionContext(issue, issueChangeContext);
+      bulkChangeData.getActionsWithoutComment().forEach(applyAction(actionContext, bulkChangeData, result));
+      addCommentIfNeeded(actionContext, bulkChangeData);
+      return result.success.contains(issue.key());
+    };
+  }
+
+  private static Consumer<Action> applyAction(ActionContext actionContext, BulkChangeData bulkChangeData, BulkChangeResult result) {
+    return action -> {
+      DefaultIssue issue = actionContext.issue();
+      try {
+        if (action.supports(issue) && action.execute(bulkChangeData.getProperties(action.key()), actionContext)) {
+          result.increaseSuccess(issue);
+        }
+      } catch (Exception e) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Either log or rethrow this exception. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1166)